glean-core/python/glean/glean.py

```diff
@@ -169,17 +179,26 @@ def configuration(cls):
         return cls._configuration
 
     @classmethod
-    def reset(cls):
+    def _reset(cls):
         """
         Resets the Glean singleton.
         """
         # TODO: 1594184 Send the metrics ping
+
+        # WARNING: Do not run any tasks on the Dispatcher from here since this
+        # is called atexit, which also waits for the Dispatcher queue to
+        # complete.
+
         Dispatcher.reset()
         if cls._initialized:
             _ffi.lib.glean_destroy_glean()
         cls._initialized = False
         if cls._destroy_data_dir and cls._data_dir.exists():
-            shutil.rmtree(str(cls._data_dir))
+            # This needs to be run in the same one-at-a-time process as the
+            # PingUploadWorker to avoid a race condition.
+            p = ProcessDispatcher.dispatch(_rmtree, (str(cls._data_dir),))
+            if not isinstance(p, bool):
```

Comment:
Ugh, different return types /o\
But I guess I'm fine with it, the alternative would be to return a type that's joinable (and throws an exception on `join` if it wraps `false`, and does nothing otherwise)

Reply:
Yeah -- I had thought about this.  It's probably clearer from the outside / messier on the inside ;)